### PR TITLE
Add table aws_s3_object

### PR DESCRIPTION
**Pull request by bigdatasourav**
# Integration test logs
<details>
  <summary>Logs</summary>

```
N/A
```
</details>

# Example query results
<details>
  <summary>Results</summary>

```
> select
  key,
  arn,
  bucket_name,
  last_modified,
  storage_class,
  version_id
from
  aws_s3_object
where
  bucket_name = 'sourav-test-bucket';
+----------------------------+------------------------------------------------------------+--------------------+---------------------------+---------------+----------------------------------+
| key                        | arn                                                        | bucket_name        | last_modified             | storage_class | version_id                       |
+----------------------------+------------------------------------------------------------+--------------------+---------------------------+---------------+----------------------------------+
| test2/                     | arn:aws:s3:::sourav-test-bucket/test2/                     | sourav-test-bucket | 2023-03-01T11:57:00+05:30 | STANDARD      | ngafgr01ri9ttfYrEgv5wArzHDLeyehS |
| test1/failed_tests (1).txt | arn:aws:s3:::sourav-test-bucket/test1/failed_tests (1).txt | sourav-test-bucket | 2023-03-01T16:30:41+05:30 | STANDARD      | mVllMkWb2T592F5ib4VndOfBY.z6T6A0 |
| test1/                     | arn:aws:s3:::sourav-test-bucket/test1/                     | sourav-test-bucket | 2023-03-01T11:56:49+05:30 | STANDARD      | pLMvvtskmhG7wmoHCklRYdSUf4kW_rom |
+----------------------------+------------------------------------------------------------+--------------------+---------------------------+---------------+----------------------------------+

> select
  key,
  arn,
  bucket_name,
  last_modified,
  storage_class,
  version_id
from
  aws_s3_object
where
  bucket_name = 'sourav-test-bucket'
  and prefix = 'test1/';
+----------------------------+------------------------------------------------------------+--------------------+---------------------------+---------------+----------------------------------+
| key                        | arn                                                        | bucket_name        | last_modified             | storage_class | version_id                       |
+----------------------------+------------------------------------------------------------+--------------------+---------------------------+---------------+----------------------------------+
| test1/                     | arn:aws:s3:::sourav-test-bucket/test1/                     | sourav-test-bucket | 2023-03-01T11:56:49+05:30 | STANDARD      | pLMvvtskmhG7wmoHCklRYdSUf4kW_rom |
| test1/failed_tests (1).txt | arn:aws:s3:::sourav-test-bucket/test1/failed_tests (1).txt | sourav-test-bucket | 2023-03-01T16:30:41+05:30 | STANDARD      | mVllMkWb2T592F5ib4VndOfBY.z6T6A0 |
+----------------------------+------------------------------------------------------------+--------------------+---------------------------+---------------+----------------------------------+

> select
  key,
  arn,
  bucket_name,
  last_modified,
  storage_class,
  version_id
from
  aws_s3_object
where
  bucket_name = 'sourav-test-bucket'
  and sse_kms_key_id is not null;
+----------------------------+------------------------------------------------------------+--------------------+---------------------------+---------------+----------------------------------+
| key                        | arn                                                        | bucket_name        | last_modified             | storage_class | version_id                       |
+----------------------------+------------------------------------------------------------+--------------------+---------------------------+---------------+----------------------------------+
| test1/failed_tests (1).txt | arn:aws:s3:::sourav-test-bucket/test1/failed_tests (1).txt | sourav-test-bucket | 2023-03-01T16:30:41+05:30 | STANDARD      | mVllMkWb2T592F5ib4VndOfBY.z6T6A0 |
+----------------------------+------------------------------------------------------------+--------------------+---------------------------+---------------+----------------------------------+
```
</details>

### CODE REVIEW

File: docs/tables/aws_s3_object.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+from
+  aws_s3_object,
+  jsonb_array_elements(aws_s3_object.acl -> 'Grants') as acl_grant
+where
+  bucket_name = 'steampipe-test'
+  and acl_grant ->> 'Permission' = 'FULL_CONTROL'
+  and acl_grant -> 'Grantee' ->> 'ID' != aws_s3_object.owner ->> 'ID';
+```
+
+### List all objects in a bucket where legal hold is on
```
Comment (misraved) [inline docs/tables/aws_s3_object.md:118]:
```suggestion
### List all objects in a bucket where the legal hold is enabled
```

File: docs/tables/aws_s3_object.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+  bucket_name,
+  object_lock_legal_hold_status
+from
+  aws_s3_object
+where
+  bucket_name = 'steampipe-test'
+  and object_lock_legal_hold_status = 'ON';
+```
+
+### List all objects in a bucket with governance lock mode
```
Comment (misraved) [inline docs/tables/aws_s3_object.md:132]:
```suggestion
### List all objects in a bucket with governance lock mode enabled
```

File: aws/table_aws_s3_object.go [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+	location, err := getBucketLocationForObjects(ctx, d, h)
+	if err != nil {
+		return nil, err
+	} else if location == "" {
+		return nil, nil
+	}
+
+	svc, err := S3Client(ctx, d, fmt.Sprint(location))
+	if err != nil {
+		plugin.Logger(ctx).Error("listS3Objects", "get_client_error", err)
```
Comment (misraved) [inline aws/table_aws_s3_object.go:377]:
```suggestion
		plugin.Logger(ctx).Error("aws_s3_object.listS3Objects", "get_client_error", err)
```
@bigdatasourav could you please append the table name to all the log statements?

File: aws/table_aws_s3_object.go [hunk 4]
```diff
@@ -0,0 +1,668 @@
...
+
+	equalQuals := d.EqualsQuals
+	if equalQuals["prefix"] != nil {
+		if equalQuals["prefix"].GetStringValue() != "" {
+			input.Prefix = aws.String(equalQuals["prefix"].GetStringValue())
+		}
+	}
+
+	// execute list call
+	objects, err := svc.ListObjectsV2(ctx, input)
```
Comment (misraved) [inline aws/table_aws_s3_object.go:409]: Does the API support pagination?

Comment (misraved) [reply aws/table_aws_s3_object.go:409]: From the API response, I see a parameter `ContinuationToken` and `NextContinuationToken` being returned, could we use it for pagination?

**Review by misraved — COMMENTED** [review]
@bigdatasourav please take a look at the review comments. Thanks!!

**Comment by misraved** [comment]
@bigdatasourav why didn't we use `listS3Buckets` as the `ParentHydrate` instead of adding `bucket_name` as the required qual?

File: docs/tables/aws_s3_object.md (hunk 1 above)
Comment (bigdatasourav) [reply docs/tables/aws_s3_object.md:118]: updated

File: docs/tables/aws_s3_object.md (hunk 2 above)
Comment (bigdatasourav) [reply docs/tables/aws_s3_object.md:132]: updated

File: aws/table_aws_s3_object.go (hunk 3 above)
Comment (bigdatasourav) [reply aws/table_aws_s3_object.go:377]: updated

File: aws/table_aws_s3_object.go [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+		return nil, err
+	}
+
+	return object, nil
+}
+
+func getS3ObjectAttributes(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	// Bucket location will be nil if getBucketLocationForObjects returned an error but
+	// was ignored through ignore_error_codes config arg
+	location, err := getBucketLocationForObjects(ctx, d, h)
```
[the remark this answers is not in the document]
Comment (bigdatasourav) [reply aws/table_aws_s3_object.go:468]: We are caching the data [here](https://github.com/turbot/steampipe-plugin-aws/pull/1623/files#diff-53d578f8338956cf4af4cb191d4f99397bdcd19476c1e77edbe7886140318923R587)

File: aws/table_aws_s3_object.go (hunk 4 above)
Comment (bigdatasourav) [reply aws/table_aws_s3_object.go]: yes, I have added the pagination

**Comment by bigdatasourav** [comment]
@misraved In a single bucket, we can have thousands of objects, so if we use parentHydrate, our execution time will be increased too much, and we will get too many throttling issues.

**Review by misraved — APPROVED** [review]

**Pull request merged**